File: paw_acp_sdk/paw_acp_sdk/directive_parser.py
"""Streaming state-machine parser for ACP directive fence syntax.

Recognises fenced directive blocks of the form::

    <<<directive
    {"type": "action_confirmation", ...}
    >>>

Everything outside those blocks is emitted as :class:`ACPTextChunk`.
"""

import json
from enum import Enum, auto
from typing import List, Optional, Set, Union

from .types import ACPTextChunk, ACPDirective


class _ACPParserState(Enum):
    STREAMING_TEXT = auto()
    MAYBE_DIRECTIVE = auto()
    IN_DIRECTIVE = auto()


class ACPDirectiveStreamParser:
    """Streaming parser that splits LLM output into text chunks and directives.

    Usage::

        parser = ACPDirectiveStreamParser(known_types={"action_confirmation", "form"})
        async for chunk in llm_stream:
            for event in parser.feed(chunk):
                if isinstance(event, ACPTextChunk):
                    # send text to user
                elif isinstance(event, ACPDirective):
                    # convert to ui.* notification
        for event in parser.flush():
            # handle remaining buffered content
    """

    _OPEN_FENCE = "<<<directive"
    _CLOSE_FENCE = ">>>"

    def __init__(self, known_types: Optional[Set[str]] = None):
        self._state = _ACPParserState.STREAMING_TEXT
        self._buffer = ""
        self._directive_body = ""
        self._fence_line = ""
        self._known_types = known_types

    def feed(self, chunk: str) -> List[Union[ACPTextChunk, ACPDirective]]:
        """Feed a chunk of text and return any events parsed so far."""
        self._buffer += chunk
        events: List[Union[ACPTextChunk, ACPDirective]] = []
        self._process(events)
        return events

    def flush(self) -> List[Union[ACPTextChunk, ACPDirective]]:
        """Call when the LLM stream is done. Flushes any buffered content."""
        events: List[Union[ACPTextChunk, ACPDirective]] = []
        if self._state == _ACPParserState.MAYBE_DIRECTIVE:
            events.append(ACPTextChunk(self._fence_line + self._buffer))
        elif self._state == _ACPParserState.IN_DIRECTIVE:
            events.append(ACPTextChunk(self._fence_line + self._directive_body + self._buffer))
        elif self._buffer:
            events.append(ACPTextChunk(self._buffer))
        self._buffer = ""
        self._reset()
        return events

    def _reset(self):
        self._state = _ACPParserState.STREAMING_TEXT
        self._directive_body = ""
        self._fence_line = ""

    def _process(self, events):
        changed = True
        while changed:
            changed = False
            if self._state == _ACPParserState.STREAMING_TEXT:
                changed = self._process_streaming_text(events)
            elif self._state == _ACPParserState.MAYBE_DIRECTIVE:
                changed = self._process_maybe_directive(events)
            elif self._state == _ACPParserState.IN_DIRECTIVE:
                changed = self._process_in_directive(events)

    def _process_streaming_text(self, events) -> bool:
        idx = self._buffer.find("<<<")
        if idx == -1:
            safe = len(self._buffer) - 2
            if safe > 0:
                events.append(ACPTextChunk(self._buffer[:safe]))
                self._buffer = self._buffer[safe:]
            return False
        if idx > 0:
            events.append(ACPTextChunk(self._buffer[:idx]))
        self._buffer = self._buffer[idx:]
        self._state = _ACPParserState.MAYBE_DIRECTIVE
        self._fence_line = ""
        return True

    def _process_maybe_directive(self, events) -> bool:
        newline_idx = self._buffer.find("\n")
        if newline_idx == -1:
            return False
        first_line = self._buffer[:newline_idx].strip()
        if first_line == self._OPEN_FENCE:
            self._fence_line = self._buffer[:newline_idx + 1]
            self._buffer = self._buffer[newline_idx + 1:]
            self._directive_body = ""
            self._state = _ACPParserState.IN_DIRECTIVE
            return True
        events.append(ACPTextChunk(self._buffer[:3]))
        self._buffer = self._buffer[3:]
        self._state = _ACPParserState.STREAMING_TEXT
        return True

    def _process_in_directive(self, events) -> bool:
        search_target = "\n" + self._CLOSE_FENCE
        close_idx = self._buffer.find(search_target)
        if close_idx == -1:
            if self._buffer.lstrip().startswith(self._CLOSE_FENCE) and self._directive_body:
                stripped = self._buffer.lstrip()
                after_fence = stripped[len(self._CLOSE_FENCE):]
                if not after_fence or after_fence[0] == '\n' or after_fence.strip() == '':
                    return self._try_parse_directive(
                        events,
                        self._directive_body,
                        self._buffer[self._buffer.index(self._CLOSE_FENCE) + len(self._CLOSE_FENCE):],
                    )
            keep = len(search_target) - 1
            safe = len(self._buffer) - keep
            if safe > 0:
                self._directive_body += self._buffer[:safe]
                self._buffer = self._buffer[safe:]
            return False

        body = self._directive_body + self._buffer[:close_idx]
        remaining = self._buffer[close_idx + len(search_target):]
        nl = remaining.find("\n")
        if nl != -1:
            remaining = remaining[nl + 1:]
        else:
            remaining = ""
        return self._try_parse_directive(events, body, remaining)

    def _try_parse_directive(self, events, body: str, remaining: str) -> bool:
        body = body.strip()
        try:
            payload = json.loads(body)
            dtype = payload.pop("type", None)
            if dtype and (self._known_types is None or dtype in self._known_types):
                events.append(ACPDirective(dtype, payload))
            else:
                events.append(ACPTextChunk(self._fence_line + body + "\n" + self._CLOSE_FENCE))
        except (json.JSONDecodeError, ValueError):
            events.append(ACPTextChunk(self._fence_line + body + "\n" + self._CLOSE_FENCE))
        self._buffer = remaining
        self._reset()
        return True

File: sdks/shepaw-acp-sdk-typescript/src/directive-parser.ts
/**
 * Streaming state-machine parser for ACP directive fence syntax.
 *
 * Recognises fenced blocks of the form:
 *
 * ```
 * <<<directive
 * {"type": "action_confirmation", ...}
 * >>>
 * ```
 *
 * Text outside those blocks is emitted as `ACPTextChunk`, and
 * well-formed fenced blocks become `ACPDirective` events.
 *
 * This is a line-for-line port of `shepaw_acp_sdk.directive_parser`
 * with identical state transitions and buffer slicing so the same
 * input reproduces the same event sequence as the Python parser.
 */

import type { ACPDirective, ACPParsedEvent, ACPTextChunk } from './types.js';

enum ParserState {
  StreamingText,
  MaybeDirective,
  InDirective,
}

const OPEN_FENCE = '<<<directive';
const CLOSE_FENCE = '>>>';

export interface ACPDirectiveStreamParserOptions {
  /**
   * When provided, only directives with a `type` in this set are emitted
   * as `ACPDirective`. Unknown types round-trip as text (preserving the
   * fence so downstream consumers can tell).
   */
  knownTypes?: ReadonlySet<string>;
}

export class ACPDirectiveStreamParser {
  private state: ParserState = ParserState.StreamingText;
  private buffer = '';
  private directiveBody = '';
  private fenceLine = '';
  private readonly knownTypes?: ReadonlySet<string>;

  constructor(opts: ACPDirectiveStreamParserOptions = {}) {
    this.knownTypes = opts.knownTypes;
  }

  /** Feed a chunk of streaming text and return any events parsed so far. */
  feed(chunk: string): ACPParsedEvent[] {
    this.buffer += chunk;
    const events: ACPParsedEvent[] = [];
    this.process(events);
    return events;
  }

  /** Flush any buffered content. Call once the upstream stream is done. */
  flush(): ACPParsedEvent[] {
    const events: ACPParsedEvent[] = [];
    if (this.state === ParserState.MaybeDirective) {
      events.push(textChunk(this.fenceLine + this.buffer));
    } else if (this.state === ParserState.InDirective) {
      events.push(textChunk(this.fenceLine + this.directiveBody + this.buffer));
    } else if (this.buffer.length > 0) {
      events.push(textChunk(this.buffer));
    }
    this.buffer = '';
    this.reset();
    return events;
  }

  // ── internal ─────────────────────────────────────────────────

  private reset(): void {
    this.state = ParserState.StreamingText;
    this.directiveBody = '';
    this.fenceLine = '';
  }

  private process(events: ACPParsedEvent[]): void {
    // Run the state machine until no further progress can be made with the
    // current buffer contents (mirrors the `changed = True` loop in Python).
    let changed = true;
    while (changed) {
      changed = false;
      if (this.state === ParserState.StreamingText) {
        changed = this.processStreamingText(events);
      } else if (this.state === ParserState.MaybeDirective) {
        changed = this.processMaybeDirective(events);
      } else if (this.state === ParserState.InDirective) {
        changed = this.processInDirective(events);
      }
    }
  }

  private processStreamingText(events: ACPParsedEvent[]): boolean {
    const idx = this.buffer.indexOf('<<<');
    if (idx === -1) {
      // No `<<<` yet — but a partial `<<` or `<` at the tail might be the
      // start of one on the next chunk, so keep the last 2 chars buffered.
      const safe = this.buffer.length - 2;
      if (safe > 0) {
        events.push(textChunk(this.buffer.slice(0, safe)));
        this.buffer = this.buffer.slice(safe);
      }
      return false;
    }
    if (idx > 0) {
      events.push(textChunk(this.buffer.slice(0, idx)));
    }
    this.buffer = this.buffer.slice(idx);
    this.state = ParserState.MaybeDirective;
    this.fenceLine = '';
    return true;
  }

  private processMaybeDirective(events: ACPParsedEvent[]): boolean {
    const newlineIdx = this.buffer.indexOf('\n');
    if (newlineIdx === -1) return false;
    const firstLine = this.buffer.slice(0, newlineIdx).trim();
    if (firstLine === OPEN_FENCE) {
      this.fenceLine = this.buffer.slice(0, newlineIdx + 1);
      this.buffer = this.buffer.slice(newlineIdx + 1);
      this.directiveBody = '';
      this.state = ParserState.InDirective;
      return true;
    }
    // Not a directive — flush the leading `<<<` as text and resume.
    events.push(textChunk(this.buffer.slice(0, 3)));
    this.buffer = this.buffer.slice(3);
    this.state = ParserState.StreamingText;
    return true;
  }

  private processInDirective(events: ACPParsedEvent[]): boolean {
    const searchTarget = '\n' + CLOSE_FENCE;
    const closeIdx = this.buffer.indexOf(searchTarget);
    if (closeIdx === -1) {
      // Allow the close fence to appear without a leading newline if the
      // buffer starts with whitespace + `>>>` (matches Python fallback).
      const stripped = this.buffer.trimStart();
      if (stripped.startsWith(CLOSE_FENCE) && this.directiveBody.length > 0) {
        const afterFence = stripped.slice(CLOSE_FENCE.length);
        if (afterFence.length === 0 || afterFence[0] === '\n' || afterFence.trim() === '') {
          const fenceStart = this.buffer.indexOf(CLOSE_FENCE);
          const remaining = this.buffer.slice(fenceStart + CLOSE_FENCE.length);
          return this.tryParseDirective(events, this.directiveBody, remaining);
        }
      }
      // Keep enough of the tail buffered to catch `\n>>>` if it straddles chunks.
      const keep = searchTarget.length - 1;
      const safe = this.buffer.length - keep;
      if (safe > 0) {
        this.directiveBody += this.buffer.slice(0, safe);
        this.buffer = this.buffer.slice(safe);
      }
      return false;
    }

    const body = this.directiveBody + this.buffer.slice(0, closeIdx);
    let remaining = this.buffer.slice(closeIdx + searchTarget.length);
    const nl = remaining.indexOf('\n');
    remaining = nl !== -1 ? remaining.slice(nl + 1) : '';
    return this.tryParseDirective(events, body, remaining);
  }

  private tryParseDirective(events: ACPParsedEvent[], bodyRaw: string, remaining: string): boolean {
    const body = bodyRaw.trim();
    try {
      const payload = JSON.parse(body) as Record<string, unknown>;
      const dtype = typeof payload.type === 'string' ? payload.type : undefined;
      // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
      delete payload.type;
      if (dtype !== undefined && (this.knownTypes === undefined || this.knownTypes.has(dtype))) {
        events.push(directiveEvent(dtype, payload));
      } else {
        events.push(textChunk(this.fenceLine + body + '\n' + CLOSE_FENCE));
      }
    } catch {
      events.push(textChunk(this.fenceLine + body + '\n' + CLOSE_FENCE));
    }
    this.buffer = remaining;
    this.reset();
    return true;
  }
}

function textChunk(content: string): ACPTextChunk {
  return { kind: 'text', content };
}

function directiveEvent(directiveType: string, payload: Record<string, unknown>): ACPDirective {
  return { kind: 'directive', directiveType, payload };
}
